diff -urN hadoop-0.20.203.0-orig/src/c++/pipes/api/hadoop/Pipes.hh hadoop-0.20.203.0/src/c++/pipes/api/hadoop/Pipes.hh
--- hadoop-0.20.203.0-orig/src/c++/pipes/api/hadoop/Pipes.hh	2011-05-04 08:30:16.000000000 +0200
+++ hadoop-0.20.203.0/src/c++/pipes/api/hadoop/Pipes.hh	2011-06-04 14:13:52.000000000 +0200
@@ -31,6 +31,8 @@
 #include <string>
 #endif
 
+#include <stdint.h>
+
 namespace HadoopPipes {
 
 /**
diff -urN hadoop-0.20.203.0-orig/src/c++/utils/api/hadoop/SerialUtils.hh hadoop-0.20.203.0/src/c++/utils/api/hadoop/SerialUtils.hh
--- hadoop-0.20.203.0-orig/src/c++/utils/api/hadoop/SerialUtils.hh	2011-05-04 08:30:16.000000000 +0200
+++ hadoop-0.20.203.0/src/c++/utils/api/hadoop/SerialUtils.hh	2011-06-04 14:09:28.000000000 +0200
@@ -19,6 +19,7 @@
 #define HADOOP_SERIAL_UTILS_HH
 
 #include <string>
+#include <stdint.h>
 
 namespace HadoopUtils {
 
