diff -ur /tmp/hadoop-0.20.1//c++/Linux-amd64-64/include/hadoop/Pipes.hh /opt/hadoop/c++/Linux-amd64-64/include/hadoop/Pipes.hh
--- /tmp/hadoop-0.20.1//c++/Linux-amd64-64/include/hadoop/Pipes.hh	2009-09-01 22:59:43.000000000 +0200
+++ /opt/hadoop/c++/Linux-amd64-64/include/hadoop/Pipes.hh	2009-11-30 16:27:24.000000000 +0100
@@ -31,6 +31,8 @@
 #include <string>
 #endif
 
+#include <stdint.h>
+
 namespace HadoopPipes {
 
 /**
diff -ur /tmp/hadoop-0.20.1//c++/Linux-amd64-64/include/hadoop/SerialUtils.hh /opt/hadoop/c++/Linux-amd64-64/include/hadoop/SerialUtils.hh
--- /tmp/hadoop-0.20.1//c++/Linux-amd64-64/include/hadoop/SerialUtils.hh	2009-09-01 22:59:43.000000000 +0200
+++ /opt/hadoop/c++/Linux-amd64-64/include/hadoop/SerialUtils.hh	2009-11-30 16:23:52.000000000 +0100
@@ -19,6 +19,7 @@
 #define HADOOP_SERIAL_UTILS_HH
 
 #include <string>
+#include <sys/types.h>
 
 namespace HadoopUtils {
 
