Index: setup.py
===================================================================
--- setup.py	(revision 1783)
+++ setup.py	(working copy)
@@ -135,6 +135,11 @@
         f = open("src/%s.cpp" % n, "w")
         f.write(contents[n])
         f.close()
+        
+    # FIXME: this is to fix the signal handling problem
+    import shutil
+    shutil.copyfile("src/HadoopPipes_signal.cpp", "src/HadoopPipes.cpp")
+    
     return contents.keys()
 
 
@@ -148,6 +153,8 @@
         ["src/%s.cpp" % n for n in aux],
         include_dirs=get_hadoop_include_dirs(HADOOP_HOME),
         libraries = ["pthread", "boost_python"],
+        define_macros=[('PYDOOP_HANDLE_SIGNALS', None),
+                       ('_POSIX_PTHREAD_SEMANTICS', None)]
         )
     return factory.create()
 
